### ISSUE TITLE
Add encrypted files support

### ISSUE THREAD

**Issue by RaphGL** [RaphGL/Tuckr#4]
Currently the Encrypts folder is simply ignored, it should use the same file structure as Hooks and Configs.

**Comment by tbmreza** [comment]
How does Encrypts folder conceptually integrate with the rest of Tuckr's workflow?

In a nutshell (please cmiiw), `Dotfiles/Configs` is basically the replacement of stow style's `Dotfiles` root, and `Dotfiles/Hooks` is where I put scripts that automatically run before/after adding a program.

I assume `Dotfiles/Encrypts` is where I can put dotenv-like files (that contains secrets like my email and passwords) that 1) Configs and Hooks can read from; 2) saved in clear text locally, like a dotenv.

Why do I need it encrypted again? And how do I manage the private/public key stuff? @RaphGL

**Comment by RaphGL** [comment]
> I assume `Dotfiles/Encrypts` is where I can put dotenv-like files (that contains secrets like my email and passwords) that 1) Configs and Hooks can read from; 2) saved in clear text locally, like a dotenv.

Yeah, Encrypts is for files that are sensitive and you might want to store somewhere but not have it exposed to everyone.
But it's not necessarily in clear text.

> Why do I need it encrypted again? And how do I manage the private/public key stuff?
You don't want people to get a hold of your API keys or your private information or any other sensitive stuff that might be needed for certain configurations. But you might still want to back it up and use it across all your devices over time.

As for how to manage it. I haven't got a particular method yet. All I know so far is that the Encrypts file will also follow the same file structure as Configs and Hooks, where you have `Encrypts/NeoMutt/somefile` and as with all the other commands, the encryption should probably be asymmetric and you use either the pre existing add a remove command or a new command specifically for encrypted files.

Then we use a similar approach to how I did with Configs (except this time it's not a symlink) and I basically try to match the `Encrypts/NeoMutt` files with the path at $HOME to give status messages.

Honestly though, I don't use encrypted files on my own dotfiles so this functionality hasn't been a priority to me and I'm not heavily opinionated about it, so I'm open to suggestions and even for people that need this for their workflow to work to make suggestions

**Comment by tbmreza** [comment]
https://github.com/AGWA/git-crypt This repo looks promising. We can port its idea to native rust, or roll a tuckr subcommand that wraps/automates it, or even just point to git-crypt documentation.

**Comment by RaphGL** [comment]
While it looks nice, it means that you would potentially have to read and understand how git crypto works and how we wrap it to be able to use this. While a small amount of configuration should be fine, I think this would probably violate the goal of the project as stated here:
> Most dotfile managers out there rely on some sort of configuration file to be able manage your dotfiles, this project came about because I couldn't find any dotfile manager that was simple enough that you could just jump into it and start using it, with no need for reading lengthy documentation and dotfile manager specific things.

Doing a quick glance at that project, it also seems to be just wrapping GPG, so maybe we could create our own set of functions around GPG that adheres to our easy to use with little to no configuration goal? (and if possible not let the windows support fall behind). We could possibly use a wrapper like [gpgme](https://github.com/gpg-rs/gpgme)?

**Comment by tbmreza** [comment]
I think we can put aside this discussion of picking which library/dependency, and design the list of side effects that our subcommands would do regarding Encrypts. When is the encryption taking place exactly? Decryption?

I'm thinking of a wizard interface:

tuckr detects that your Encrypts folder is not empty. Do you want to setup gpg for tuckr, encrypt files in the Encrypts folder, git stage them, and add the original unencrypted version to .gitignore?
> [Y/n]
(select public key or whatever)
...

**Comment by RaphGL** [comment]
> When is the encryption taking place exactly? Decryption?
I think we should strive for consistency so I was thinking we could "overload" the add and remove commands and let it act on both Encrypts and Configs.

> I'm thinking of a wizard interface
Yeah that's a good idea. 

> Do you want to setup gpg for tuckr, encrypt files in the Encrypts folder, git stage them, and add the original unencrypted version to .gitignore?
I think we should setup gpg for the Encrypts folder and add their unencrypted equivalents to gitignore yeah. This will also allow us to just leverage the Symlink functionality to it, so we can just symlink the unencrypted files into home and just have all the symlink management commands work with them. 
Then we could supply some additional commands for encryption specific stuff if need be? 

Let me know what you think.

**Comment by tbmreza** [comment]
> I think we should strive for consistency so I was thinking we could "overload" the add and remove commands and let it act on both Encrypts and Configs.

Got it.

> This will also allow us to just leverage the Symlink functionality to it, so we can just symlink the unencrypted files into home and just have all the symlink management commands work with them.

Something just clicked in my head. I was confusing between (maybe we need to support both) parts of a dotfile are secret, and a whole dotfile is secret.

The second case is easy. For the first case, let's assume a dotfile .frob that contains:
```
signin $EMAIL_PSWD
```

and a .frob.env:
```
EMAIL_PSWD=hunter12
```

This second case is useful for when you don't want to manually sync the version you can share publicly and the one that's encrypted.

**Comment by RaphGL** [comment]
So you're suggesting that we add a partial secret type to Encrypts? 
If I'm not wrong the alternative you're suggesting is providing a sort of diff? If so should we exclude the sensitive parts instead of everything else? That way we could make so that if there's a diff we only treat that part of the file as secret and if there's `-- ` at the start of lines or something.
```diff
-- signin $EMAIL_PSWD
-- darkmode dimmed
-- max_logfile_size 50
....
basically a bunch of configs that are not sensitive
```

I also wouldn't oppose changing the Encrypts folder to something like Secrets if it stops being strictly encrypted files.

**Comment by tbmreza** [comment]
> So you're suggesting that we add a partial secret type to Encrypts?

Not necessarily, just something to keep in mind, that there can be 2 valid usecases of encrypted dotfile (whole and partial). To be honest I'm not sure which one does appeal to more people.

Now that you mentioned Secrets folder, I think we can avoid doing encryption/gpg entirely:

Git ignore its contents, hydrate Configs with its secret variables and symlink the resulting file (also git ignore it). Wdyt?

**Comment by RaphGL** [comment]
I'm sorry but I didn't quite understand what you meant in the last comment
what did you mean with 
>  hydrate Configs with its secret variables

**Comment by tbmreza** [comment]
search (for $ variables) and replace (with its value)

**Comment by RaphGL** [comment]
oh that, I don't know if that's a great idea. That would bloat the user's environment with a bunch of variables and they'd need to keep track of them somehow and some secrets are entire files. Also, if said used wanted to back up these variables where would they put them? on their .profile or .zshenv or something? That would means that their entire environment setup is sensitive and cannot be backed up.

**Comment by tbmreza** [comment]
That's a criticism for the partially sensitive dotfiles usecase, if I understand things correctly. But I agree with the criticism.


Do we agree that 1) when we say dotfiles that contain sensitive data, we mean the whole file is encrypted? And 2) tuckr makes it easy, automatic when possible, to add to .config and share (via git repo) such dotfiles, managing gpg under the hood or otherwise?

We are near to the conclusion of this discussion yay 😀

**Comment by RaphGL** [comment]
> 1) when we say dotfiles that contain sensitive data, we mean the whole file is encrypted? 
Yeah, if need be we can just add partial encryption later on.

> 2) tuckr makes it easy, automatic when possible, to add to .config and share (via git repo) such dotfiles, managing gpg under the hood or otherwise?
Yeah, automatic should be default unless it's not feasible.

**Comment by RaphGL** [comment]
Hi, I'm going to be working on this feature soon as the only thing missing for this project to be fully functional is adding encryption and testing more on windows (as I don't really use windows often). Let me know if you're still interesting in working on this issue @tbmreza

**Comment by tbmreza** [comment]
I'm sorry but since that last comment that I made I've been swamped with my day job + something else ☹️. Please proceed with your plan 🙇‍♂️. @RaphGL

**Comment by RaphGL** [comment]
@tbmreza no problem, feel free to contribute anytime if u want :)
